api/catalog/api/controllers/elasticsearch/search.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import annotations
+
+import json
+import logging as log
+import pprint
+from typing import List, Literal, Tuple, Union
+
+from django.conf import settings
+
+from elasticsearch.exceptions import RequestError
+from elasticsearch_dsl import Q, Search
+from elasticsearch_dsl.response import Hit
+
+from catalog.api.controllers.elasticsearch.utils import (
+    _exclude_filtered_providers,
+    _get_query_slice,
+    _get_result_and_page_count,
+    _post_process_results,
+)
+from catalog.api.serializers.media_serializers import MediaSearchRequestSerializer
+
+
+def _quote_escape(query_string: str) -> str:
+    """
+    If there are any unmatched quotes in the query supplied by the user, ignore
+    them by escaping.
+
+    :param query_string: the string in which to escape unbalanced quotes
+    :return: the given string, if the quotes are balanced, the escaped string otherwise
+    """
+
+    num_quotes = query_string.count('"')
+    if num_quotes % 2 == 1:
+        return query_string.replace('"', '\\"')
+    else:
+        return query_string
+
+
+def _apply_filter(
+    s: Search,
+    query_ser: MediaSearchRequestSerializer,
+    basis: Union[str, tuple[str, str]],
+    behaviour: Literal["filter", "exclude"] = "filter",
+) -> Search:
+    """
+    Parse and apply a filter from the search parameters serializer. The
+    parameter key is assumed to have the same name as the corresponding
+    Elasticsearch property. Each parameter value is assumed to be a comma
+    separated list encoded as a string.
+
+    :param s: the search query to issue to Elasticsearch
+    :param query_ser: the ``MediaSearchRequestSerializer`` instance with search query
+    :param basis: the name of the field in the serializer and Elasticsearch
+    :param behaviour: whether to accept (``filter``) or reject (``exclude``) the hit
+    :return: the modified search query
+    """
+
+    search_params = query_ser.data
+    if isinstance(basis, tuple):
+        serializer_field, es_field = basis
+    else:
+        serializer_field = es_field = basis
+    if serializer_field in search_params:
+        filters = []
+        for arg in search_params[serializer_field].split(","):
+            filters.append(Q("term", **{es_field: arg}))
+        method = getattr(s, behaviour)  # can be ``s.filter`` or ``s.exclude``
+        return method("bool", should=filters)
```

Comment:
From my limited understanding of ES, I _think_ we can replace the `should` with `must` here, because we don't need the results that do not contain the filter parameter (i.e., when we are filtering by some value, we don't need the score that `should` can bring).
This is not a suggestion for this PR, I'm just putting my thoughts in words here :)

Reply:
I thought so too, but I don't know enough ES to be confident enough to make the change. I'll update it if you're sure that it should be `must` instead of `should`.